lib/discover_endpoint.js

```diff
@@ -0,0 +1,340 @@
+  var service = request.service;
+  var api = service.api || {};
+  var operations = api.operations;
+  var identifiers = {};
+  if (operations[request.operation] && operations[request.operation].name) {
+    identifiers.operation = operations[request.operation].name;
+  }
+  if (service.config.region) {
+    identifiers.region = service.config.region;
+  }
+  if (api.serviceId) {
+    identifiers.serviceId = api.serviceId
+  }
+  if (service.config.credentials.accessKeyId) {
+    identifiers.accessKeyId = service.config.credentials.accessKeyId
+  }
+  return identifiers;
+}
+
+/**
+ * @api private
+ */
+function marshallCustomIdentifiersHelper(result, params, shape) {
+  if (!shape || params === undefined || params === null) return;
+  if (shape.type === 'structure' && shape.required && shape.required.length > 0) {
+    util.arrayEach(shape.required, function(name) {
+      var memberShape = shape.members[name];
+      if (memberShape.endpointDiscoveryId === true) {
+        var locationName = memberShape.isLocationName ? memberShape.name : name;
+        result[locationName] = String(params[name]);
+      } else {
+        marshallCustomIdentifiersHelper(result, params[name], memberShape);
+      }
+    });
+  }
+}
+
+/**
+ * Get customized cache key according to the 'endpointDiscoveryId' trait.
+ * @param [object] request object
+ * @param [object] input shape of the given operation's api
+ */
+function marshallCustomIdentifiers(request, shape) {
+  var identifiers = {};
+  marshallCustomIdentifiersHelper(identifiers, request.params, shape);
+  return identifiers;
+}
+
+/**
+ * Call endpoint discovery operation when it's optional.
+ * When endpoint is available in cache then use the cached endpoints. If endpoints
+ * are unavailable then use regional endpoints and call endpoint discovery operation
+ * asynchronously. This is turned off by default.
+ * @param [object] request object
+ * @api private
+ */
+function optionalDiscoverEndpoint(request) {
+  var service = request.service;
+  var api = service.api;
+  var operationModel = api.operations ? api.operations[request.operation] : undefined;
+  var inputShape = operationModel ? operationModel.input : undefined;
+
+  var identifiers = marshallCustomIdentifiers(request, inputShape);
+  var cacheKey = getCacheKey(request);
+  cacheKey = util.update(cacheKey, identifiers);
+  var endpoints = AWS.endpointCache.get(cacheKey);
+  if (endpoints && endpoints.length === 1 && endpoints[0].Address === '') {
+    //endpoint operation is being made but response not yet received
+    //or endpoint operation just failed in 1 minute
+    return;
+  } else if (endpoints && endpoints.length > 0) {
+    //found endpoint record from cache
+    request.httpRequest.updateEndpoint(endpoints[0].Address);
+  } else {
+    //endpoint record not in cache or outdated. make discovery operation
+    var endpointRequest = service.makeRequest(api.endpointOperation, {
+      Operation: operationModel.name,
```

Comment:
It looks like the request structure for the endpoint operation can be optionally supplied. If the operation does not model `Operation` or `Identifiers`, should they still be included? Do you know if it will cause problems to send those fields if the service doesn't expect them?

Reply:
`Operation` and `Identifiers` are always supplied. If the endpoint operation input shape doesn't include any of them, marshaller will exclude them anyway.

Reply:
Sweet, thanks for verifying.